packages/transaction-messages/src/use-address-lookup-tables.ts

```diff
@@ -0,0 +1,86 @@
+import { Address } from '@solana/addresses';
+import { AccountRole, IAccountLookupMeta, IInstruction, isSignerRole } from '@solana/instructions';
+
+import { AddressesByLookupTableAddress } from './addresses-by-lookup-table-address';
+import { ITransactionMessageWithFeePayer } from './fee-payer';
+import { TransactionMessage } from './transaction-message';
+
+type Mutable<T> = {
+    -readonly [P in keyof T]: T[P];
+};
+
+// Look up the address in lookup tables, return a lookup meta if it is found in any of them
+function findAddressInLookupTables(
+    address: Address,
+    role: AccountRole.READONLY | AccountRole.WRITABLE,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): IAccountLookupMeta | undefined {
+    for (const [lookupTableAddress, addresses] of Object.entries(addressesByLookupTableAddress)) {
+        for (let i = 0; i < addresses.length; i++) {
+            if (address === addresses[i]) {
+                return {
+                    address,
+                    addressIndex: i,
+                    lookupTableAddress: lookupTableAddress as Address,
+                    role,
+                };
+            }
+        }
+    }
+}
+
+type TransactionMessageNotLegacy = Exclude<TransactionMessage, { version: 'legacy' }>;
+
+export function useAddressLookupTablesForTransactionMessage<
+    TTransactionMessage extends TransactionMessageNotLegacy = TransactionMessageNotLegacy,
+>(
+    transactionMessage: TTransactionMessage,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): TTransactionMessage {
+    // All signers will need to be static keys, so we shouldn't replace them if they're used elsewhere in the transaction,
+    // even if they're also in a lookup table
+    const signers = new Set(
+        transactionMessage.instructions
+            .flatMap(i => (i.accounts ?? []).filter(a => isSignerRole(a.role)))
+            .map(a => a.address),
+    );
+    if ('feePayer' in transactionMessage) {
+        signers.add((transactionMessage as ITransactionMessageWithFeePayer).feePayer);
+    }
+
+    const lookupTableAddresses = new Set(Object.values(addressesByLookupTableAddress).flatMap(a => a));
+
+    const newInstructions: IInstruction[] = [];
```

Comment:
You want ‘if nothing changed, return the exact same object’ behaviour here. Like [this](https://github.com/solana-labs/solana-web3.js/blob/9239e6ec972b4de9f0d15b197fbef1d2871759d9/packages/transactions/src/signatures.ts#L82-L84), tested [here](https://github.com/solana-labs/solana-web3.js/blob/9239e6ec972b4de9f0d15b197fbef1d2871759d9/packages/transactions/src/__tests__/signatures-test.ts#L257-L266).